### ISSUE TITLE
Incremental preview not working with `Subs` command

### ISSUE THREAD

**Issue by akinsho** [johmsalas/text-case.nvim#13]
Hi @johmsalas,

Not sure which commit introduced the bug, but I'm no longer seeing a preview when I use the `Subs` command. It doesn't matter what mode I'm in. I'm on the latest commit of the repo. Sorry this is a bit brief, I'm hoping it's easily reproducible and not specific to my system or config.

**Comment by johmsalas** [comment]
@akinsho I'm seeing this behavior on nvim with Plug

I have 2 setups.

**1. Lunarvim + Lua using Packer**. There is this initialization method https://github.com/johmsalas/text-case.nvim/blob/6e69c13089e70388a359fe63c4ef92053f7fcbcb/lua/textcase/plugin/presets.lua#L98
It works fine

**2. nvim raw + Plug, with this setup**
This fails
```
call plug#begin('~/.local/share/nvim/plugged')

Plug 'johmsalas/text-case.nvim', {'branch': 'issue-6_provide-key-mapping-desc'}

call plug#end()

nnoremap gau :lua require('textcase').current_word('to_upper_case')<CR> 
nnoremap geu :lua require('textcase').operator('to_upper_case')<CR>
```

I think in the second case. [Initialize](https://github.com/johmsalas/text-case.nvim/blob/6e69c13089e70388a359fe63c4ef92053f7fcbcb/lua/textcase/plugin/presets.lua#L98) is not called. But I'm not a vim-plug user, I don't really know if there is some kind of lazy loading

------------------------

Bottom line. I think it stopped working for you when the [registration of the command](https://github.com/johmsalas/text-case.nvim/blob/6e69c13089e70388a359fe63c4ef92053f7fcbcb/lua/textcase/plugin/presets.lua#L20) was moved from the M.setup method, to the Initialize method

**Comment by akinsho** [comment]
@johmsalas I'm also a packer user so not sure if that's a thing that would be the difference, lunarvim is quite big so can easily imagine there are all sorts of loading logic in there somewhere. OOI why move the logic out of setup and also importantly you use a global function with a pretty common name so I can easily imagine that it could clash with another plugin using a global function with the same name. I think in general its quite unsafe to use globals like that in nvim since all plugins can pollute the global namespace.

Anyway, it seems that just requiring the `presets` file should have triggered the `Initialise` call, but it doesn't for whatever reason on my side. I just called it manually and everything then works as expected.

**Comment by johmsalas** [comment]
Actually, the initialize should be working with Plug, because the key maps works. Glad it works for you. I'm going to have a look and debug it

@akinsho What strategy do  you use for releasing in your pluggin? This pluggin will soon reach "stable" and I'm thinking about unit testing. But also some branching release? tags? do you have some experience with that?

**Comment by akinsho** [comment]
>  Glad it works for you
Just to be clear it's not actually working automatically the `Initialise` function is declared but not called, I think it should not be global because that's a bug waiting to happen, it will just take some other plugin author creating a plugin that uses `Initialise` as well and 💥 .

Anyway, re. testing I use plenary's busted testing framework and that works well for me. It might be quite hard to assert things like position of highlights in your plugin that way though, but you could test all the intermediate logic 🤷🏿

**Comment by johmsalas** [comment]
Sorry, weekend... But, yes, that could be a local func. Learning LUA, sorry. I'm changing it to a local var

The reason for moving the function to its own auto called method, is because some people prefer to define their own keybindings. My rational was: Initialize basic things, finish the setup with customizable options

I found it, https://www.linode.com/docs/guides/writing-a-neovim-plugin-with-lua/:

This would be the source of the bug:
> Lua files are not sourced until they are called on via the require function

And this the solution:
> A directory named plugin that defines the plugin. This directory should have a Vim script (.vim) file using your plugin’s name. The file defines the commands that the plugin should expose and sets up any keybindings that you want the plugin to have by default. It is also the only Vim script code required for your Neovim plug

That's why it works once the file is required.

--------
Btw, Thanks for your answer about the testing framework. This one is the experiment to learn LUA and vim plugins. I'm working on another plugin and want to learn as much as possible so that one is successful.

I can assert the position of highlights by decoupling those methods and testing them isolated

**Comment by johmsalas** [comment]
Should be working now https://github.com/johmsalas/text-case.nvim/pull/15/files
Thanks for reporting :)

**Comment by johmsalas** [comment]
By the way, haven't update the readme but you could find this useful:

```
vim.api.nvim_set_keymap('n', 'gae', "<cmd>TextCaseStartReplacingCommand<CR>", { desc = "Replace text" })
vim.api.nvim_set_keymap('v', 'gae', "<cmd>TextCaseStartReplacingCommand<CR>", { desc = "Replace text" })
```

They start the `Subs` command and prefills the original text with whatever is under the cursor or visually selected

**Comment by akinsho** [comment]
Thanks, @johmsalas that works now, but (there's always a but 😅) I've noticed that when removing the word it doesn't do any highlighting, it only highlights the replacement e.g.
`:Subs/apple` at this point apple is not highlighted at all, then when you press `:Subs/apple/pear` you get a preview for pear. So you don't see apple disappear till you press `/` then start typing pear and that works

**Comment by johmsalas** [comment]
Hi!

This was reported as an individual issue and solved [here](https://github.com/johmsalas/text-case.nvim/issues/24) .

Going to solve this one
